pageserver/src/repository.rs

```diff
@@ -0,0 +1,131 @@
+//pub mod rocksdb;
+pub mod inmemory;
+
+use bytes::{Bytes, BytesMut, Buf, BufMut};
+use crate::ZTimelineId;
+use zenith_utils::lsn::Lsn;
+use crate::waldecoder::Oid;
+use std::sync::Arc;
+
+use anyhow::Result;
+
+//
+// A repository corresponds to one .zenith directory. One repository holds multiple
+// timelines, forked off from the same initial call to 'initdb'.
+//
+pub trait Repository {
+    // FIXME: I wish these would return an abstract `&dyn Timeline` instead
+    fn get_timeline(&self, timelineid: ZTimelineId) -> Result<Arc<inmemory::InMemoryTimeline>>;
+    fn get_or_restore_timeline(&self, timelineid: ZTimelineId) -> Result<Arc<inmemory::InMemoryTimeline>>;
+
+    //fn get_stats(&self) -> RepositoryStats;
+}
+
+pub trait Timeline {
+
+    // Look up given page in the cache.
+    fn get_page_at_lsn(&self, tag: BufferTag, lsn: Lsn) -> Result<Bytes>;
+    fn get_relsize(&self, tag: RelTag, lsn: Lsn) -> Result<u32>;
+    fn get_relsize_exists(&self, tag: RelTag, lsn: Lsn) -> Result<bool>;
+
+    // Functions used by WAL receiver
+    fn put_wal_record(&self, tag: BufferTag, rec: WALRecord);
+    fn put_truncation(&self, tag: RelTag, lsn: Lsn, relsize: u32) -> anyhow::Result<()>;
+    fn put_page_image(&self, tag: BufferTag, lsn: Lsn, img: Bytes);
+    fn create_database(&self, lsn: Lsn, db_id: Oid, tablespace_id: Oid, src_db_id: Oid, src_tablespace_id: Oid) -> Result<()>;
+
+    fn advance_last_record_lsn(&self, lsn: Lsn);
+    fn advance_last_valid_lsn(&self, lsn: Lsn);
+    fn init_valid_lsn(&self, lsn: Lsn);
+    fn get_last_valid_lsn(&self) -> Lsn;
+}
```

Comment:
Rustdoc comments on the traits and trait member fns would be a nice gift to new people reading the code.

It would be nice if we could do this for all public items (structs, enums, fns, traits...)